python/tank_vendor/shotgun_authentication/session_cache.py

```diff
@@ -0,0 +1,209 @@
+    """
+    # Retrieve the location of the cached info
+    info_path = _get_session_data_location(base_url)
+    # Nothing was cached, return an empty dictionary.
+    if not os.path.exists(info_path):
+        logger.debug("No cache found at %s" % info_path)
+        return None
+    try:
+        # Read the login information
+        config = SafeConfigParser({"login": None, "session_token": None})
+        config.read(info_path)
+        if not config.has_section("LoginInfo"):
+            logger.debug("No Login info was found")
+            return None
+
+        login = config.get("LoginInfo", "login", raw=True)
+        session_token = config.get("LoginInfo", "session_token", raw=True)
+
+        if not login or not session_token:
+            logger.debug("Incomplete settings (login:%s, session_token:%s)" % (login, session_token))
+            return None
+
+        return {"login": login, "session_token": session_token}
+    except Exception:
+        logger.exception("Exception thrown while loading cached session info.")
+        return None
+
+
+def cache_session_data(host, login, session_token):
+    """
+    Caches the session data for a site and a user.
+
+    :param host: Site we want to cache a session for.
+    :param login: User we want to cache a session for.
+    :param session_token: Session token we want to cache.
+    """
+    # Retrieve the cached info file location from the host
+    info_path = _get_session_data_location(host)
+
+    # make sure the info_dir exists!
+    info_dir, info_file = os.path.split(info_path)
+    if not os.path.exists(info_dir):
+        os.makedirs(info_dir, 0700)
+
+    logger.debug("Caching login info at %s...", info_path)
+    # Create a document with the following format:
+    # [LoginInfo]
+    # login=username
+    # session_token=some_unique_id
+    config = SafeConfigParser()
+    config.add_section("LoginInfo")
+    config.set("LoginInfo", "login", login)
+    config.set("LoginInfo", "session_token", session_token)
+    # Write it to disk.
+    with open(info_path, "w") as configfile:
+        config.write(configfile)
+    logger.debug("Cached!")
+
+
+def generate_session_token(hostname, login, password, http_proxy):
+    """
+    Generates a session token for a given username/password on a given site.
+
+    :param hostname: The host to connect to.
+    :param login: The user to get a session for.
+    :param password: Password for the user.
+    :param http_proxy: Proxy to use. Can be None.
+
+    :returns: The generated session token for that user/password/site combo.
+
+    :raises: AuthenticationError if the credentials were invalid.
+    """
+    try:
+        # Create the instance...
+        sg = Shotgun(
+            hostname,
+            login=login,
+            password=password,
+            http_proxy=http_proxy
+        )
+        # .. and generate the session token. If it throws, we have invalid credentials.
+        return sg.get_session_token()
```

Comment:
am i right thinking this will do two rountrips to shotgun, one for the connect and one for the session fetch? I think we should try setting the `connect` constructor parameter in the instantiation call to False and see if that helps improve performance ( https://github.com/shotgunsoftware/python-api/blob/master/shotgun_api3/shotgun.py#L270 ). This is a non-frequent operation, so not a big deal, but since it's potentially such a quick fix, I thought it's worth a try!